src/main.cpp

```diff
@@ -4502,6 +4534,33 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
     }
 
 
+    // Disconnect existing peer connection if:
+    // 1. The version message has been received
+    // 2. Overwinter is active
+    // 3. Peer version is legacy (pre-Overwinter)
+    // 4. Peer branch id is not supported
+    // FIXME: one branch id or set of branch ids?
+    else if (
+        pfrom->nVersion != 0 &&
+        NetworkUpgradeActive(GetHeight(), chainparams.GetConsensus(), Consensus::UPGRADE_TESTDUMMY))
+    {
+        if (pfrom->nVersion < OVERWINTER_PROTO_VERSION) {
+            LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, pfrom->nVersion);
+            pfrom->PushMessage("reject", strCommand, REJECT_OBSOLETE,
+                               strprintf("Version must be %d or greater", OVERWINTER_PROTO_VERSION));
+            pfrom->fDisconnect = true;
+            return false;
+        }
+        else if (pfrom->nBranchId != NetworkUpgradeInfo[Consensus::UPGRADE_TESTDUMMY].nBranchId) {
```

Comment:
If `nBranchId` ends up being the branch ID of the peer's chain tip at time of connect, we need to think about how it gets updated (new network messages?)

If it ends up being the branch ID the peer will eventually target, I think this should be fine, because to change that, the peer will need to drop its connection and restart.